index.js

```diff
@@ -0,0 +1,114 @@
+import assert from 'node:assert'
+import ms from 'ms'
+
+export class CancelStuckTransactions {
+  #store
+  #list
+  #resolve
+  #log
+  #sendTransaction
+  constructor ({ store, list, resolve, log, sendTransaction }) {
+    this.#store = store
+    this.#list = list
+    this.#resolve = resolve
+    this.#log = log
+    this.#sendTransaction = sendTransaction
+  }
+
+  async pending (tx) {
+    assert.strictEqual(typeof tx.hash, 'string')
+    assert.strictEqual(typeof tx.from, 'string')
+    assert.strictEqual(typeof tx.maxPriorityFeePerGas, 'bigint')
+    assert.strictEqual(typeof tx.nonce, 'number')
+    await this.#store({
+      ...tx,
+      timestamp: new Date().toISOString()
+    })
+  }
+
+  async successful (tx) {
+    assert.strictEqual(typeof tx.hash, 'string')
+    await this.#resolve(tx.hash)
+  }
+
+  async olderThan (ageMs) {
+    assert.strictEqual(typeof ageMs, 'number')
+
+    this.#log('Checking for stuck transactions...')
+    const txs = await this.#list()
+    const txsToCancel = txs.filter(tx => {
+      return new Date() - new Date(tx.timestamp) > ageMs
+    })
+    if (txsToCancel.length === 0) {
+      this.#log('No transactions to cancel')
+      return
+    }
+
+    this.#log('Transactions to cancel:')
+    for (const tx of txsToCancel) {
+      this.#log(
+        `- ${tx.hash} (age ${ms(new Date() - new Date(tx.timestamp))})`
+      )
+    }
+
+    const recentSendMessage = await this.#getRecentSendMessage()
+    this.#log(
+      'Calculating gas fees from the recent Send message ' +
+      recentSendMessage.cid +
+      ' (created at ' +
+      new Date(recentSendMessage.timestamp * 1000).toISOString() +
+      ')'
+    )
+
+    await Promise.all(txsToCancel.map(tx => this.#cancelTx({
+      tx,
+      recentGasUsed: recentSendMessage.receipt.gasUsed,
+      recentGasFeeCap: Number(recentSendMessage.gasFeeCap)
+    })))
+  }
+
+  async #cancelTx ({ tx, recentGasUsed, recentGasFeeCap }) {
+    // Increase by 25% + 1 attoFIL (easier: 25.2%) and round up
+    const maxPriorityFeePerGas = (tx.maxPriorityFeePerGas * 1252n + 1000n) / 1000n
+    const gasLimit = Math.ceil(recentGasUsed * 1.1)
+
+    this.#log(`Replacing ${tx.hash}...`)
+    this.#log(`- maxPriorityFeePerGas: ${tx.maxPriorityFeePerGas} -> ${maxPriorityFeePerGas}`)
+    this.#log(`- gasLimit: ${recentGasUsed} -> ${gasLimit}`)
+    const replacementTx = await this.#sendTransaction({
+      to: tx.from,
+      value: 0,
+      nonce: tx.nonce,
+      gasLimit,
+      maxFeePerGas: maxPriorityFeePerGas > recentGasFeeCap
+        ? maxPriorityFeePerGas
+        : recentGasFeeCap,
+      maxPriorityFeePerGas
+    })
+    this.#log(`
+      Waiting for receipt of replacing ${tx.hash} with ${replacementTx.hash}...`
+    )
+    await replacementTx.wait()
```

Comment:
Please add error handling.

- `sendTransaction` can fail with `err.code === 'NONCE_EXPIRED'` if we are too late and the transaction was added to a recently mined block
- `replacementTx.wait()` throws an error if the TX was rejected, I don't have a good example to show, but I encountered this while running my auto-cancellation script

It would be great to extract `cancelTx` and `getRecentSendMessage` into a public function exported by this module, so that I can re-use it in my script.

https://github.com/filecoin-station/utils/pull/1

Reply:
> replacementTx.wait() throws an error if the TX was rejected, I don't have a good example to show, but I encountered this while running my auto-cancellation script

What kind of error handling should I add for this though? 🤔

Reply:
> I guess we have to treat this error as a generic "unknown error" and let the caller figure out what to do about it (e.g. report the problem to Sentry).

That's how I was thinking about it as well 👍 

> Is it worth mentioning in the docs that cancelOlderThan() will throw if the network rejects some of the replacement transactions?

Since we don't know yet how it can throw, I'd advertise for letting it through and then documenting these errors.

I'll also document the known nonce error next.

> Also, since we are cancelling multiple TXs concurrently, we need to ensure the error allows us to link the error details to the hash of the TX we were waiting to get confirmed. Linking to both the replaced and the replacement TXs will be even better.

When you run `cancelOlderThan()` it does a `Promise.all()` of canceling all transactions. The first one to reject will create the error that is thrown. I don't know yet whether the tx will be attached already. Are you ok with letting it fail and then seeing how to best improve?

Reply:
> > Also, since we are cancelling multiple TXs concurrently, we need to ensure the error allows us to link the error details to the hash of the TX we were waiting to get confirmed. Linking to both the replaced and the replacement TXs will be even better.
> 
> When you run `cancelOlderThan()` it does a `Promise.all()` of canceling all transactions. The first one to reject will create the error that is thrown.

Hmm, what happens with the other concurrent tasks when one of them fails and `Promise.all` returns? Will the tasks continue to completion until the replacement TX is submitted and confirmed? What happens if the caller calls `cancelOlderThan` again before those tasks are complete? Can we end up with multiple replacement transactions for the same stuck transaction?

Maybe we should use `Promise.allSettled` instead.

> I don't know yet whether the tx will be attached already. Are you ok with letting it fail and then seeing how to best improve?

Sure 👍🏻

Reply:
> Maybe we should use Promise.allSettled instead.

I was thinking about this as well. In the old implementation (before this lib), individual cancel failures would simply be logged (and retried later) but the overall operation would succeed. I'll do as advised

Reply:
https://github.com/filecoin-station/cancel-stuck-transactions/pull/1/commits/b24f50a3623306f848c38c4aaf9957684bb1232f